### PR TITLE
upgrade go to 1.16 and add SAN to all certs

### LINKED ISSUE #1899: Migrate to using SANs for webhook certificates for Go 1.15
Context: https://github.com/googleforgames/agones/pull/1897

Go 1.15 no longer supports Common Name based certs for HTTPS without a specific flag. Common Name is a 20 year old deprecation, and we probably shouldn't be promoting.

There is a backward compatible environment flag, but it will be going away eventually. This flag, for the life of me I could not make it work -- which blocked a migration to latest 1.15.

I tried to write steps to create certs with SANs, but my OpenSSL knowledge is not good enough.

So, as a plan to migrate to Go 1.15:

1. [x] Update [webhook with cert creation](https://agones.dev/site/docs/getting-started/create-webhook-fleetautoscaler/#2-create-x509-root-and-webhook-certificates) instructions to use SANs over Common Name, and ensure it works with current Agones version.
2. [x] Provide release note announcement on deprecation of Common Name certificates for webhooks.
3. [ ] Migrate build system to Go 1.15 and no longer support SAN based certificates.

### LINKED ISSUE #2024: Can't run Allocation example with Go 1.16
**What happened**:

Attempting to run the allocation service sample with Go 1.16 results in the following error:

```
$ go run main.go --ip 35.232.217.29 --namespace default --key /home/markmandel/cloud-game-servers-examples/terraform-4-cluster/allocation/certs/gke_gcgs-beta-demo_us-central1-b_game-cluster-us-2/client.key --cert /home/markmandel/cloud-game-servers-examples/terraform-4-cluster/allocation/certs/gke_gcgs-beta-demo_us-central1-b_game-cluster-us-2/client.crt --cacert /home/markmandel/cloud-game-servers-examples/terraform-4-cluster/allocation/certs/gke_gcgs-beta-demo_us-central1-b_game-cluster-us-2/ca.crt --multicluster truepanic: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: authentication handshake failed: x509: cannot validate certificate for 35.232.217.29 because it doesn't contain any IP SANs"
goroutine 1 [running]:main.main()        /home/markmandel/agones/examples/allocator-client/main.go:76 +0x81cexit status 2
$ go version
go version go1.16 linux/amd64
```

**What you expected to happen**:

It to successfully allocate.

**How to reproduce it (as minimally and precisely as possible)**:

Follow https://agones.dev/site/docs/advanced/allocator-service/ with openssl certificates, and then attempt to allocate.

**Anything else we need to know?**:

Not sure of workaround at this stage.

**Environment**:
- Agones version: 1.12.0
- Kubernetes version (use `kubectl version`): 1.17.0
- Cloud provider or hardware configuration: GKE
- Install method (yaml/helm): Helm
- Troubleshooting guide log(s): Above
- Others: N/A

### PR DESCRIPTION
**Pull request by cindy52**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug

kind cleanup
> /kind documentation
> /kind feature
> /kind hotfix

**What this PR does / Why we need it**:
Upgrade go version to the latest (1.16.5). Go 1.15 no longer supports Common Name based certs for HTTPS without a specific flag. This PR includes adding SAN for all related certs and updating the doc to create a cert with SAN for allocation service.

**Which issue(s) this PR fixes**:

Closes #1899 
Closes #2024 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by markmandel** [comment]
In case you didn't see it:
```
docker run -e FILE="/go/src/agones.dev/agones/site/content/en/docs/Reference/agones_crd_api_reference.html" -e VERSION=1.16.0 --rm -i -v /workspace/build//.config/gcloud:/root/.config/gcloud -v ~/.kube/:/root/.kube -v ~/.config/helm:/root/.config/helm -v ~/.cache/helm:/root/.cache/helm -v /workspace:/go/src/agones.dev/agones -v /workspace/build//.gomod:/go/pkg/mod -v /workspace/build//.gocache:/root/.cache/go-build agones-build:d2646eda2f bash -c "/go/src/agones.dev/agones/site/gen-api-docs.sh"
using go proxy as a workaround for git.agache.org being down: http://proxy.golang.org
includes/website.mk:100: recipe for target 'test-gen-api-docs' failed
make[1]: Leaving directory '/workspace/build'
```

That seems weird actually, I'm not sure what is going on there. Can you repro locally?

Usually that means need to run `gen-api-docs`, but I don't see a diff issue there 🤔

**Comment by cindy52** [comment]
> In case you didn't see it:
> 
> ```
> docker run -e FILE="/go/src/agones.dev/agones/site/content/en/docs/Reference/agones_crd_api_reference.html" -e VERSION=1.16.0 --rm -i -v /workspace/build//.config/gcloud:/root/.config/gcloud -v ~/.kube/:/root/.kube -v ~/.config/helm:/root/.config/helm -v ~/.cache/helm:/root/.cache/helm -v /workspace:/go/src/agones.dev/agones -v /workspace/build//.gomod:/go/pkg/mod -v /workspace/build//.gocache:/root/.cache/go-build agones-build:d2646eda2f bash -c "/go/src/agones.dev/agones/site/gen-api-docs.sh"
> using go proxy as a workaround for git.agache.org being down: http://proxy.golang.org
> includes/website.mk:100: recipe for target 'test-gen-api-docs' failed
> make[1]: Leaving directory '/workspace/build'
> ```
> 
> That seems weird actually, I'm not sure what is going on there. Can you repro locally?
> 
> Usually that means need to run `gen-api-docs`, but I don't see a diff issue there 🤔

It's weird. Every time it fails with different errors. Last time I checked it failed with `build-agones-sdk-image-windows`....

**Comment by cindy52** [comment]
`gen-api-docs` currently failed because In Go 1.16, module-aware commands report an error after discovering a problem in go.mod or go.sum instead of attempting to fix the problem automatically. The related issue can be found here https://github.com/golang/go/issues/44529. I tried to fix it as suggested by changing the replacement and add `-mod=mod` with `go build` in `gen-api-docs.sh` file. The build was successfully after that but I'm getting an error `type invalid type has kind=Unsupported which is unhandled` when generating the api doc. Found some related issues here https://github.com/kubeflow/website/pull/1899. Looks like there is not a known fix yet.

File: install/helm/agones/certs/cert.sh [hunk 1]
```diff
@@ -4,7 +4,7 @@ echo "Generating certs..."
 echo "Email should be: emailAddress=agones-discuss@googlegroups.com"
 echo "Common Name should be: agones-controller-service.agones-system.svc"
 openssl genrsa -out server.key 2048
-openssl req -new -x509 -sha256 -key server.key -out server.crt -days 3650
+openssl req -new -x509 -sha256 -key server.key -out server.crt -days 3650 -addext 'subjectAltName=DNS:agones-controller-service.agones-system.svc'
```
Comment (markmandel) [inline install/helm/agones/certs/cert.sh:7]: Aaah. That's how you do it!

**Comment by markmandel** [comment]
Total shot in the dark, but this is the library we use:
https://github.com/ahmetb/gen-crd-api-reference-docs

I see a similar issue here: https://github.com/ahmetb/gen-crd-api-reference-docs/issues/10

Also wondering if an upgrade helps? 🤷🏻

**Comment by cindy52** [comment]
> Total shot in the dark, but this is the library we use:
> https://github.com/ahmetb/gen-crd-api-reference-docs
> 
> I see a similar issue here: [ahmetb/gen-crd-api-reference-docs#10](https://github.com/ahmetb/gen-crd-api-reference-docs/issues/10)
> 
> Also wondering if an upgrade helps? 🤷🏻

I tried upgrading to the latest but still got the same err..

**Comment by markmandel** [comment]
Digging into how other people fixed this on their end, looks like they found a solution here: https://github.com/knative/docs/pull/2054/files

But the author does work here at Google (he's on my team 😄 ) we could always go bug him and see what he knows.

It makes me feel like we're missing a reference to something somewhere.

**Comment by cindy52** [comment]
> Digging into how other people fixed this on their end, looks like they found a solution here: https://github.com/knative/docs/pull/2054/files
> 
> But the author does work here at Google (he's on my team 😄 ) we could always go bug him and see what he knows.
> 
> It makes me feel like we're missing a reference to something somewhere.

I saw it yesterday and even looks into his error logs in the issue: https://github.com/ahmetb/gen-crd-api-reference-docs/issues/10. I believe the cause is different. For knative issue, it turns out some of the import paths are github.com  instead of knative.dev, it shows in the error log he attached. I turned on detailed log as well and all import paths are agones.dev. It shows in the details of the failed build.

**Comment by markmandel** [comment]
Should we close this PR for now?

**Comment by cindy52** [comment]
> Should we close this PR for now?

Done!

**Pull request closed without merging**